File: convex/approvals.ts
import { mutation } from "./_generated/server";
import { v } from "convex/values";

export const createApproval = mutation({
  args: { taskId: v.string(), action: v.any(), status: v.string() },
  handler: async ({ db }, { taskId, action, status }) => {
    const createdAt = Date.now();
    const _id = await db.insert("approvals", { taskId, action, status, createdAt });
    return { ok: true, _id };
  },
});

export const approveTask = mutation({
  args: { taskId: v.string() },
  handler: async ({ db }, { taskId }) => {
    const latest = await db
      .query("approvals")
      .withIndex("by_taskId", (q) => q.eq("taskId", taskId))
      .collect();
    if (!latest.length) return { error: "not_found" };
    // Assume last action is the one to approve
    latest.sort((a, b) => b.createdAt - a.createdAt);
    const approval = latest[0];
    await db.insert("approvals", {
      taskId,
      action: approval.action,
      status: "approved",
      createdAt: Date.now(),
    });
    return { taskId, action: approval.action, status: "approved" };
  },
});

export const rejectTask = mutation({
  args: { taskId: v.string() },
  handler: async ({ db }, { taskId }) => {
    const latest = await db
      .query("approvals")
      .withIndex("by_taskId", (q) => q.eq("taskId", taskId))
      .collect();
    if (!latest.length) return { error: "not_found" };
    latest.sort((a, b) => b.createdAt - a.createdAt);
    const approval = latest[0];
    await db.insert("approvals", {
      taskId,
      action: approval.action,
      status: "rejected",
      createdAt: Date.now(),
    });
    return { taskId, action: approval.action, status: "rejected" };
  },
});

import { query } from "./_generated/server";

export const getByTask = query({
  args: { taskId: v.string() },
  handler: async ({ db }, { taskId }) => {
    const items = await db
      .query("approvals")
      .withIndex("by_taskId", (q) => q.eq("taskId", taskId))
      .collect();
    items.sort((a, b) => a.createdAt - b.createdAt);
    return { approvals: items };
  },
});

File: backend/core/approvals.py
from backend.core.events import emit_event
from backend.core.convex_client import convex
import asyncio

async def require_approval(task_id, action: dict):
    # Store approval in Convex
    await convex.mutation("approvals:createApproval", {
        "taskId": task_id,
        "action": action,
        "status": "pending"
    })
    
    await emit_event(task_id, "pending_approval", {"action": action})
    return {"status": "waiting_for_approval"}

async def approve_task(task_id):
    # Call Convex mutation to approve task
    result = await convex.mutation("approvals:approveTask", {"taskId": task_id})
    
    if result.get("error"):
        return {"status": "not_found"}
    
    action = result.get("action", {})
    await emit_event(task_id, "approved", {"action": action})
    
    # Continue workflow after approval
    await continue_workflow_after_approval(task_id, action)
    
    return {"status": "approved", "action": action}

async def reject_task(task_id):
    # Call Convex mutation to reject task
    result = await convex.mutation("approvals:rejectTask", {"taskId": task_id})
    
    if result.get("error"):
        return {"status": "not_found"}
    
    action = result.get("action", {})
    await emit_event(task_id, "rejected", {"action": action})
    return {"status": "rejected"}

async def continue_workflow_after_approval(task_id, action):
    """Continue workflow execution after approval"""
    from backend.core.payments import process_payment
    from backend.core.memory import fetch_policy
    
    try:
        # Step 4 – Process payment (after approval)
        payment = await process_payment(task_id, action["amount"], action["recipient"])
        
        # Get policy for completion event
        policy = await fetch_policy(task_id, "contractor commission")
        
        await emit_event(task_id, "completed", {"policy": policy, "payment": payment})
    except Exception as e:
        await emit_event(task_id, "failed", {"error": str(e)})
